pasture-algorithms/src/normal_estimation.rs

```diff
@@ -0,0 +1,551 @@
+    rows.push(scaled_matrix.row(1).cross(&scaled_matrix.row(2)));
+
+    let mut cross_product = DMatrix::<f64>::zeros(3, 3);
+
+    // write rows of cross product
+    for it in cross_product.row_iter_mut().zip(rows) {
+        let (mut cross_row, row) = it;
+        // row from rows vector is written to the row of the cross product
+        cross_row.copy_from(&row);
+    }
+
+    // find largest eigen vector
+    let mut largest_eigen_vec = cross_product.row(0);
+    for row in cross_product.row_iter() {
+        if row.norm() > largest_eigen_vec.norm() {
+            largest_eigen_vec = row;
+        }
+    }
+
+    // set eigen vector to largest vector
+    for i in 0..eigen_vector.len() {
+        eigen_vector[i] = largest_eigen_vec[i];
+    }
+}
+
+/// for a given 3x3 matrix the functions calculates the eigen vector of the smallest eigen value that can be found
+fn eigen_3x3(covariance_matrix: &DMatrix<f64>, eigen_value: &mut f64, eigen_vector: &mut Vec<f64>) {
+    let scale = covariance_matrix.abs().max();
+    let mut covariance_matrix_scaled = DMatrix::<f64>::zeros(3, 3);
+    for i in 0..covariance_matrix.len() {
+        covariance_matrix_scaled[i] = covariance_matrix[i] / scale;
+    }
+
+    // scale the matrix down
+    for (index, value) in covariance_matrix.iter().enumerate() {
+        covariance_matrix_scaled[index] = value / scale;
+    }
+
+    let mut eigen_values = vec![0.0; 3];
+    solve_polynomial(covariance_matrix, &mut eigen_values);
+    // undo scale for smallest eigen vector
+    *eigen_value = eigen_values[0] * scale;
+
+    // subtract the smallest eigen value from the diagonal of the matrix
+    covariance_matrix_scaled
+        .diagonal()
+        .iter_mut()
+        .for_each(|x| *x = *x - eigen_values[0]);
+
+    get_largest_eigen_vector(&mut covariance_matrix_scaled, eigen_vector);
+}
+
+/// calculates the orientation of the surface as a normal vector with components n_x, n_y and n_z as well as the curvature of the surface for a given covariance matrix
+fn solve_plane_parameter(
+    covariance_matrix: &DMatrix<f64>,
+    eigen_vector: &mut Vec<f64>,
+    curvature: &mut f64,
+) {
+    let mut eigen_value: f64 = 0.0;
+    eigen_3x3(covariance_matrix, &mut eigen_value, eigen_vector);
+
+    let eigen_sum = covariance_matrix[0] + covariance_matrix[4] + covariance_matrix[8];
+    if eigen_sum != 0.0 {
+        *curvature = (eigen_value / eigen_sum).abs();
+    } else {
+        *curvature = 0.0;
+    }
+}
+
+/// calculates the normal vectors and the curvature of the surface for the given point cloud
+fn normal_estimation<T: PointBuffer>(
+    //this needs to be k nearest neighbors
+    point_cloud: &T,
+    n_x: &mut f64,
+    n_y: &mut f64,
+    n_z: &mut f64,
+    curvature: &mut f64,
+) {
+    let mut covariance_matrix = DMatrix::<f64>::zeros(3, 3);
+
+    if point_cloud.len() < 3 || compute_covarianz_matrix(point_cloud, &mut covariance_matrix) == 0 {
```

Comment:
`point_cloud.len() >= 3` is an invariant of the function that can be easily checked, so you should `panic` in this case (this is a user error). 
An invalid covariance matrix is a runtime error that can occur depending on the structure of the points. In this case, you can either return `Option<...>` from the function, or, even better, return `Result<...>` and describe what kind of error caused the failure (e.g. 'could not compute covariance matrix due to XYZ')

Reply:
I'm not sure how I should implement the Option/Result remark as the validity of the points with which the covariance matrix is calculated is checked within the function

Reply:
You can have the method return `Result<...>` and whenever an error occurs, e.g. because you can't calculate a valid covariance matrix, you simply early-exit with `Err(...)` instead of doing a `panic!`. Here is some pseudo-code, using the `anyhow` `Result` type:

```
fn foo<T: PointBuffer>(pc: &T) -> Result<()> {
  if pc.len() < 3 {
    panic!("Require at least 3 points");
  }
  let covariance_matrix = compute_covariance_matrix(pc).ok_or(anyhow!("Could not calculate covariance matrix"))?;
  //Do stuff...
  Ok()
}

// Which requires that compute_covariance_matrix returns Option<DMatrix<f64>>:

fn compute_covariance_matrix<T: PointBuffer>(pc: &T) -> Option<DMatrix<f64>> {
  //...
}
```